### PR TITLE
storage/engine: benchmark MVCCDeleteRange

### PR DESCRIPTION
**Pull request by tamird**
Motivated by #3103. cc @petermattis @mrtracy 

```
$ go test ./storage/engine -run - -bench MVCCDeleteRange 2>/dev/null
PASS
BenchmarkMVCCDeleteRange1Version8Bytes-4          10     190795633 ns/op       2.75 MB/s
BenchmarkMVCCDeleteRange1Version32Bytes-4         10     128830121 ns/op       4.07 MB/s
BenchmarkMVCCDeleteRange1Version256Bytes-4        50      38724836 ns/op      13.54 MB/s
ok      github.com/cockroachdb/cockroach/storage/engine 13.159s

$ go test ./storage/engine -run - -bench MVCCDeleteRange -memprofilerate=1 -memprofile mem.out 2>/dev/null
...
$ go tool pprof --alloc_objects engine.test mem.out
Entering interactive mode (type "help" for commands)
(pprof) top10
934147 of 1136965 total (82.16%)
Dropped 142 nodes (cum <= 5684)
Showing top 10 nodes out of 34 (cum >= 534539)
      flat  flat%   sum%        cum   cum%
    133418 11.73% 11.73%     133418 11.73%  github.com/cockroachdb/cockroach/roachpb.(*Value).Unmarshal
    133418 11.73% 23.47%     133418 11.73%  github.com/cockroachdb/cockroach/storage/engine.(*MVCCMetadata).Marshal
    133418 11.73% 35.20%     133418 11.73%  github.com/cockroachdb/cockroach/storage/engine.(*MVCCValue).Marshal
    133418 11.73% 46.94%     333545 29.34%  github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).getProtoInternal
     66913  5.89% 52.82%     401833 35.34%  github.com/cockroachdb/cockroach/storage/engine.setupMVCCData
     66726  5.87% 58.69%      66726  5.87%  github.com/cockroachdb/cockroach/storage/engine.mvccEncodeTimestamp
     66709  5.87% 64.56%      66744  5.87%  github.com/cockroachdb/cockroach/roachpb.(*Value).InitChecksum
     66709  5.87% 70.43%     200127 17.60%  github.com/cockroachdb/cockroach/storage/engine.(*MVCCValue).Unmarshal
     66709  5.87% 76.29%      66709  5.87%  github.com/cockroachdb/cockroach/storage/engine.(*rocksDBBatch).GetProto
     66709  5.87% 82.16%     534539 47.01%  github.com/cockroachdb/cockroach/storage/engine.MVCCIterate
```

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3105)

### CODE REVIEW

File: storage/engine/rocksdb_test.go [hunk 1]
```diff
@@ -484,6 +494,40 @@ func BenchmarkMVCCMergeTimeSeries(b *testing.B) {
...
+	const rangeBytes = 512 * 1024
+	const overhead = 48 // Per key/value overhead (empirically determined)
+	numKeys := rangeBytes / (overhead + valueBytes)
+
+	b.SetBytes(rangeBytes)
+	b.ResetTimer()
+	b.StopTimer()
+
+	for i := 0; i < b.N; i++ {
+		rocksdb, stopper := setupMVCCData(1, numKeys, valueBytes, false, b)
```
Comment (petermattis) [inline storage/engine/rocksdb_test.go]: Continually recreating the data is likely to be slow. I think you get rid of the changes to `setupMVCCData` and instead copy the directory and its contents on each loop iteration.

**Comment by petermattis** [comment]
`MVCCDeleteRange` is currently implemented as `MVCCScan` followed by `MVCCDelete`. I imagine there is a big win from integrating the scan with the delete.

**Comment by petermattis** [comment]
LGTM, though do you mind waiting to merge until #3096 goes in (just needs to pass tests)? I'd also be interested to know if #3096 affects this benchmark (it should). Btw, when you compare old and new, run each benchmark 10 times (`go test -count 10`) and use `benchstat` to compare the output.

**Comment by tamird** [comment]
Sure, I can wait. Not sure what you mean about old and new, since there is only one result set here.

**Comment by petermattis** [comment]
Old = before #3096 is merged.
New = after #3096 is merged.

File: storage/engine/rocksdb_test.go (hunk 1 above)
Comment (tamird) [inline storage/engine/rocksdb_test.go]: Done.

File: storage/engine/rocksdb_test.go [hunk 2]
```diff
@@ -136,51 +138,61 @@ func TestRocksDBCompaction(t *testing.T) {
...
 		}
-		if err := batch.Commit(); err != nil {
-			b.Fatal(err)
-		}
-		batch.Close()
+		rocksdb.CompactRange(nil, nil)
+		stopper.Stop()
+	}
+
+	locDirty := loc + "_dirty"
```
Comment (petermattis) [inline storage/engine/rocksdb_test.go]: I was thinking you would only make the copy in `BenchmarkMVCCDeleteRange`. Performing the copy for every benchmark seems excessive.

**Comment by tamird** [comment]
Results of rebasing on #3096:
```
name                               old time/op    new time/op    delta
MVCCDeleteRange1Version8Bytes-4       138ms ± 0%     132ms ± 1%  -4.79%   (p=0.000 n=8+10)
MVCCDeleteRange1Version32Bytes-4     98.1ms ± 1%    93.4ms ± 1%  -4.79%  (p=0.000 n=10+10)
MVCCDeleteRange1Version256Bytes-4    27.2ms ± 1%    25.9ms ± 1%  -4.77%   (p=0.000 n=10+9)

name                               old speed      new speed      delta
MVCCDeleteRange1Version8Bytes-4    3.80MB/s ± 1%  3.98MB/s ± 1%  +4.75%   (p=0.000 n=10+8)
MVCCDeleteRange1Version32Bytes-4   5.34MB/s ± 1%  5.61MB/s ± 1%  +5.01%  (p=0.000 n=10+10)
MVCCDeleteRange1Version256Bytes-4  19.3MB/s ± 1%  20.2MB/s ± 1%  +5.00%   (p=0.000 n=10+9)
```

File: storage/engine/rocksdb_test.go (hunk 2 above)
Comment (tamird) [inline storage/engine/rocksdb_test.go]: done

**Comment by tschottdorf** [comment]
LGTM


---

Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[storage/engine/rocksdb_test.go, line 511 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3105#-K2wO8wJoPf78O9ti1uC)** ([raw file](https://github.com/cockroachdb/cockroach/blob/0ca1a54ea9f7e6de720d14f3eb4544edf616733a/storage/engine/rocksdb_test.go#L511)):</sup>
does setting those attributes make any difference?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3105)

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[storage/engine/rocksdb_test.go, line 511 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3105#-K2wO8wJoPf78O9ti1uC)** ([raw file](https://github.com/cockroachdb/cockroach/blob/0ca1a54ea9f7e6de720d14f3eb4544edf616733a/storage/engine/rocksdb_test.go#L511)):</sup>
No, they do nothing. Looks like a historical accident that these attributes are on the engine rather than the store. Removed.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3105)

**Pull request merged**